File: 1497-design-a-stack-with-increment-operation/design-a-stack-with-increment-operation.py
class CustomStack:

    def __init__(self, maxSize: int):
        self.size = maxSize
        self.stack = []
        self.increment_list = []

    def push(self, x: int) -> None:
        if len(self.stack) < self.size:
            self.stack.append(x)
            self.increment_list.append(0)

    def pop(self) -> int:
        if not self.stack:
            return -1
        if len(self.increment_list) > 1:
            self.increment_list[-2] += self.increment_list[-1]
        return self.stack.pop() + self.increment_list.pop()        

    def increment(self, k: int, val: int) -> None:
        if self.increment_list:
            self.increment_list[min(k-1, len(self.stack) - 1)] += val


# Your CustomStack object will be instantiated and called as such:
# obj = CustomStack(maxSize)
# obj.push(x)
# param_2 = obj.pop()
# obj.increment(k,val)

File: 1497-design-a-stack-with-increment-operation/design-a-stack-with-increment-operation.cpp
class CustomStack {
    vector<int> stack;
    vector<int> increment_list;
    int size;
    int count = 0;
    int diff = 0;

public:
    CustomStack(int maxSize) {
        this->size = maxSize;
        this->stack.resize(this->size);
        this->increment_list.resize(this->size);
    }
    
    void push(int x) {
        if (count == size)
            return;

        if (count >= 1)
            increment_list[count - 1] += diff;
        diff = 0;

        stack[count] = x;
        increment_list[count] = 0;
        count++;
    }
    
    int pop() {
        if (count == 0)
            return -1;

        diff += increment_list[count-1];
        int result = stack[count-1] + diff;
        count--;
        return result;
    }
    
    void increment(int k, int val) {
        if (count == 0)
            return;
        increment_list[min(k - 1, count - 1)] += val;
    }
};
/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack* obj = new CustomStack(maxSize);
 * obj->push(x);
 * int param_2 = obj->pop();
 * obj->increment(k,val);
 */
